From ed3978adbc7e063568e4f51b575d34edd3a8c1b8 Mon Sep 17 00:00:00 2001
From: Sergey Shelukhin <sershe@apache.org>
Date: Wed, 11 Jan 2017 13:05:59 -0800
Subject: [PATCH 0908/1164] CDH-48968 HIVE-15551 : memory leak in directsql
 for mysql+bonecp specific initialization (Xiaomin
 Zhang, reviewed by Sergey Shelukhin)

(cherry picked from commit 2d813f4d4a0bb42345d153c362f7416f05ab2749)

Change-Id: I66ac959c45a94ca9ef76aa7345410fb2fd7976cd
---
 .../hadoop/hive/metastore/MetaStoreDirectSql.java  |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
index 089ed4a..7db18c4 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
@@ -23,6 +23,7 @@
 
 import com.google.common.collect.Lists;
 import java.sql.Connection;
+import java.sql.Statement;
 import java.sql.SQLException;
 import java.text.ParseException;
 import java.util.ArrayList;
@@ -260,12 +261,17 @@ private void doDbSpecificInitializationsBeforeQuery() throws MetaException {
 
   private void executeNoResult(final String queryText) throws SQLException {
     JDOConnection jdoConn = pm.getDataStoreConnection();
+    Statement statement = null;
     boolean doTrace = LOG.isDebugEnabled();
     try {
       long start = doTrace ? System.nanoTime() : 0;
-      ((Connection)jdoConn.getNativeConnection()).createStatement().execute(queryText);
+      statement = ((Connection)jdoConn.getNativeConnection()).createStatement();
+      statement.execute(queryText);
       timingTrace(doTrace, queryText, start, doTrace ? System.nanoTime() : 0);
     } finally {
+      if(statement != null){
+          statement.close();
+      }
       jdoConn.close(); // We must release the connection before we call other pm methods.
     }
   }
-- 
1.7.9.5

